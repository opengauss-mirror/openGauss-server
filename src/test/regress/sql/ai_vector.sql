create database test;
\c test
CREATE TABLE dimens3_scan_l2_100 (
    id serial PRIMARY KEY,
    name text,
    embedding vector(3)  -- 假设向量的维度为512
);
CREATE INDEX ON dimens3_scan_l2_100 USING hnsw (embedding vector_l2_ops);
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item1', '[0.211557005, 0.076130312, 0.048887434]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item2', '[0.822932576, 0.487093015, 0.748730227]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item3', '[0.564357002, 0.926498683, 0.818280197]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item4', '[0.791316587, 0.460895557, 0.067207831]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item5', '[0.974898792, 0.371204436, 0.090201541]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item6', '[0.972023039, 0.902695732, 0.104805143]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item7', '[0.340851511, 0.661115819, 0.979039013]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item8', '[0.492086968, 0.671430133, 0.715509519]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item9', '[0.815762744, 0.638665527, 0.940596042]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item10', '[0.598264407, 0.626925064, 0.837062887]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item11', '[0.605249576, 0.62827664, 0.776847171]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item12', '[0.608682007, 0.608462876, 0.816056557]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item13', '[0.361436461, 0.552522446, 0.010793276]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item14', '[0.856674254, 0.395471228, 0.988986555]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item15', '[0.694758094, 0.875162429, 0.232631982]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item16', '[0.021518759, 0.565872631, 0.643064199]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item17', '[0.766803278, 0.164139074, 0.516602078]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item18', '[0.764914268, 0.523333259, 0.449667669]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item19', '[0.212920548, 0.985840962, 0.093416858]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item20', '[0.789230643, 0.564697507, 0.134702261]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item21', '[0.717458865, 0.267429064, 0.178979577]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item22', '[0.509153041, 0.641964123, 0.920561135]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item23', '[0.431243764, 0.963082203, 0.727807302]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item24', '[0.827596123, 0.134412498, 0.004161737]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item25', '[0.258203395, 0.659945327, 0.597459666]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item26', '[0.944380157, 0.702332123, 0.545305512]');
set hnsw_ef_search=10;
analyze dimens3_scan_l2_100;
SELECT * FROM dimens3_scan_l2_100 ORDER BY embedding <-> '[3,1,2]' LIMIT 5;
SELECT * FROM dimens3_scan_l2_100 ORDER BY embedding <-> '[3,1,2]' LIMIT 20;
SELECT * FROM dimens3_scan_l2_100 ORDER BY embedding <-> '[3,1,2]';
explain (verbose on, costs off) SELECT /*+ indexscan(dimens3_scan_l2_100 dimens3_scan_l2_100_embedding_idx) */ * FROM dimens3_scan_l2_100 ORDER BY embedding <-> '[3,1,2]' LIMIT 5;
SELECT /*+ indexscan(dimens3_scan_l2_100 dimens3_scan_l2_100_embedding_idx) */ * FROM dimens3_scan_l2_100 ORDER BY embedding <-> '[3,1,2]' LIMIT 5;
explain (verbose on, costs off) SELECT /*+ indexscan(dimens3_scan_l2_100 dimens3_scan_l2_100_embedding_idx) */ * FROM dimens3_scan_l2_100 ORDER BY embedding <-> '[3,1,2]';
SELECT /*+ indexscan(dimens3_scan_l2_100 dimens3_scan_l2_100_embedding_idx) */ * FROM dimens3_scan_l2_100 ORDER BY embedding <-> '[3,1,2]';
prepare p_select as SELECT /*+ indexscan(dimens3_scan_l2_100 dimens3_scan_l2_100_embedding_idx) */ * FROM dimens3_scan_l2_100 ORDER BY embedding <-> $1 LIMIT $2;
execute p_select('[3,1,2]', 5);
prepare p_select1 as SELECT /*+ indexscan(dimens3_scan_l2_100 dimens3_scan_l2_100_embedding_idx) */ * FROM dimens3_scan_l2_100 ORDER BY embedding <-> '[3,1,2]' LIMIT $1;
execute p_select1(5);
prepare p_select2 as SELECT /*+ indexscan(dimens3_scan_l2_100 dimens3_scan_l2_100_embedding_idx) */ * FROM dimens3_scan_l2_100 ORDER BY embedding <-> $1 LIMIT 5;
execute p_select2('[3,1,2]');

CREATE TABLE items (id int, embedding vector(3));
INSERT INTO items VALUES (1, '[1,2,3]'), (2, '[4,5,6]'), (3, '[7,8,9]'), (4, '[10,11,12]'), (5, '[13,14,15]');
CREATE INDEX ON items USING ivfflat (embedding vector_l2_ops) WITH (lists = 100);
set enable_seqscan=off;
SELECT /*+ indexscan(embedding, items_embedding_idx) */embedding FROM items ORDER BY embedding <-> '[3,1,2]' LIMIT 5;
set ivfflat_probes = 3;
SELECT /*+ indexscan(embedding, items_embedding_idx) */embedding FROM items ORDER BY embedding <-> '[3,1,2]' LIMIT 5;

CREATE TABLE t1 (val vector(4));
INSERT INTO t1 (val) VALUES ('[0,0,0,0]'), ('[1,2,3,0]'), ('[1,1,1,0]');
INSERT INTO t1 (val) VALUES ('[4,0,0,0]'), ('[4,2,3,0]'), ('[4,1,1,0]');
INSERT INTO t1 (val) VALUES ('[25,20,20,22]'), ('[25,22,23,26]'), ('[25,21,21,21]');
INSERT INTO t1 (val) VALUES ('[24,20,20,20]'), ('[24,22,23,20]'), ('[24,21,21,20]');
CREATE INDEX ON t1 USING ivfflat (val vector_l2_ops) WITH (lists=2);
set enable_seqscan=off;
set ivfflat_probes = 1;
SELECT * FROM t1 ORDER BY val <-> '[1,2,3,4]';
\c regression
drop database test;
