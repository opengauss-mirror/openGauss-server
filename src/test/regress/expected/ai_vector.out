create database test;
\c test
CREATE TABLE dimens3_scan_l2_100 (
    id serial PRIMARY KEY,
    name text,
    embedding vector(3)  -- 假设向量的维度为512
);
NOTICE:  CREATE TABLE will create implicit sequence "dimens3_scan_l2_100_id_seq" for serial column "dimens3_scan_l2_100.id"
NOTICE:  CREATE TABLE / PRIMARY KEY will create implicit index "dimens3_scan_l2_100_pkey" for table "dimens3_scan_l2_100"
CREATE INDEX ON dimens3_scan_l2_100 USING hnsw (embedding vector_l2_ops);
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item1', '[0.211557005, 0.076130312, 0.048887434]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item2', '[0.822932576, 0.487093015, 0.748730227]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item3', '[0.564357002, 0.926498683, 0.818280197]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item4', '[0.791316587, 0.460895557, 0.067207831]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item5', '[0.974898792, 0.371204436, 0.090201541]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item6', '[0.972023039, 0.902695732, 0.104805143]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item7', '[0.340851511, 0.661115819, 0.979039013]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item8', '[0.492086968, 0.671430133, 0.715509519]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item9', '[0.815762744, 0.638665527, 0.940596042]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item10', '[0.598264407, 0.626925064, 0.837062887]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item11', '[0.605249576, 0.62827664, 0.776847171]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item12', '[0.608682007, 0.608462876, 0.816056557]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item13', '[0.361436461, 0.552522446, 0.010793276]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item14', '[0.856674254, 0.395471228, 0.988986555]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item15', '[0.694758094, 0.875162429, 0.232631982]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item16', '[0.021518759, 0.565872631, 0.643064199]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item17', '[0.766803278, 0.164139074, 0.516602078]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item18', '[0.764914268, 0.523333259, 0.449667669]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item19', '[0.212920548, 0.985840962, 0.093416858]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item20', '[0.789230643, 0.564697507, 0.134702261]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item21', '[0.717458865, 0.267429064, 0.178979577]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item22', '[0.509153041, 0.641964123, 0.920561135]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item23', '[0.431243764, 0.963082203, 0.727807302]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item24', '[0.827596123, 0.134412498, 0.004161737]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item25', '[0.258203395, 0.659945327, 0.597459666]');
INSERT INTO dimens3_scan_l2_100 (name, embedding) VALUES ('item26', '[0.944380157, 0.702332123, 0.545305512]');
set hnsw_ef_search=10;
analyze dimens3_scan_l2_100;
SELECT * FROM dimens3_scan_l2_100 ORDER BY embedding <-> '[3,1,2]' LIMIT 5;
 id |  name  |             embedding              
----+--------+------------------------------------
 14 | item14 | [0.85667425,0.39547122,0.98898655]
  9 | item9  | [0.81576276,0.6386655,0.94059604]
 26 | item26 | [0.94438016,0.70233214,0.5453055]
  2 | item2  | [0.8229326,0.487093,0.74873024]
 10 | item10 | [0.5982644,0.62692505,0.8370629]
(5 rows)

SELECT * FROM dimens3_scan_l2_100 ORDER BY embedding <-> '[3,1,2]' LIMIT 20;
 id |  name  |             embedding              
----+--------+------------------------------------
 14 | item14 | [0.85667425,0.39547122,0.98898655]
  9 | item9  | [0.81576276,0.6386655,0.94059604]
 26 | item26 | [0.94438016,0.70233214,0.5453055]
  2 | item2  | [0.8229326,0.487093,0.74873024]
 10 | item10 | [0.5982644,0.62692505,0.8370629]
 12 | item12 | [0.60868204,0.60846287,0.81605655]
  3 | item3  | [0.564357,0.9264987,0.8182802]
 11 | item11 | [0.6052496,0.62827665,0.7768472]
 22 | item22 | [0.50915307,0.64196414,0.92056113]
 18 | item18 | [0.7649143,0.52333325,0.44966766]
  6 | item6  | [0.972023,0.9026957,0.10480514]
 17 | item17 | [0.76680326,0.16413908,0.5166021]
  8 | item8  | [0.49208698,0.6714301,0.71550953]
  5 | item5  | [0.9748988,0.37120444,0.09020154]
 23 | item23 | [0.43124378,0.9630822,0.7278073]
  7 | item7  | [0.34085152,0.6611158,0.979039]
 15 | item15 | [0.6947581,0.8751624,0.23263198]
 20 | item20 | [0.78923064,0.5646975,0.13470227]
  4 | item4  | [0.79131657,0.46089557,0.06720783]
 21 | item21 | [0.71745884,0.26742905,0.17897958]
(20 rows)

SELECT * FROM dimens3_scan_l2_100 ORDER BY embedding <-> '[3,1,2]';
 id |  name  |             embedding              
----+--------+------------------------------------
 14 | item14 | [0.85667425,0.39547122,0.98898655]
  9 | item9  | [0.81576276,0.6386655,0.94059604]
 26 | item26 | [0.94438016,0.70233214,0.5453055]
  2 | item2  | [0.8229326,0.487093,0.74873024]
 10 | item10 | [0.5982644,0.62692505,0.8370629]
 12 | item12 | [0.60868204,0.60846287,0.81605655]
  3 | item3  | [0.564357,0.9264987,0.8182802]
 11 | item11 | [0.6052496,0.62827665,0.7768472]
 22 | item22 | [0.50915307,0.64196414,0.92056113]
 18 | item18 | [0.7649143,0.52333325,0.44966766]
  6 | item6  | [0.972023,0.9026957,0.10480514]
 17 | item17 | [0.76680326,0.16413908,0.5166021]
  8 | item8  | [0.49208698,0.6714301,0.71550953]
  5 | item5  | [0.9748988,0.37120444,0.09020154]
 23 | item23 | [0.43124378,0.9630822,0.7278073]
  7 | item7  | [0.34085152,0.6611158,0.979039]
 15 | item15 | [0.6947581,0.8751624,0.23263198]
 20 | item20 | [0.78923064,0.5646975,0.13470227]
  4 | item4  | [0.79131657,0.46089557,0.06720783]
 21 | item21 | [0.71745884,0.26742905,0.17897958]
 24 | item24 | [0.8275961,0.1344125,0.004161737]
 25 | item25 | [0.2582034,0.6599453,0.5974597]
 16 | item16 | [0.02151876,0.5658726,0.6430642]
 13 | item13 | [0.36143646,0.5525224,0.010793276]
 19 | item19 | [0.21292055,0.985841,0.093416855]
  1 | item1  | [0.211557,0.076130316,0.048887435]
(26 rows)

CREATE TABLE items (id int, embedding vector(3));
INSERT INTO items VALUES (1, '[1,2,3]'), (2, '[4,5,6]'), (3, '[7,8,9]'), (4, '[10,11,12]'), (5, '[13,14,15]');
CREATE INDEX ON items USING ivfflat (embedding vector_l2_ops) WITH (lists = 100);
NOTICE:  ivfflat index created with little data
DETAIL:  This will cause low recall.
HINT:  Drop the index until the table has more data.
set enable_seqscan=off;
SELECT /*+ indexscan(embedding, items_embedding_idx) */embedding FROM items ORDER BY embedding <-> '[3,1,2]' LIMIT 5;
 embedding  
------------
 [13,14,15]
 [10,11,12]
 [7,8,9]
 [4,5,6]
 [1,2,3]
(5 rows)

set ivfflat_probes = 3;
SELECT /*+ indexscan(embedding, items_embedding_idx) */embedding FROM items ORDER BY embedding <-> '[3,1,2]' LIMIT 5;
 embedding  
------------
 [13,14,15]
 [10,11,12]
 [7,8,9]
 [4,5,6]
 [1,2,3]
(5 rows)

\c regression
drop database test;
